### ISSUE TITLE
Using micromamba in a bash script

### ISSUE THREAD

**Issue by mil-ad** [mamba-org/mamba#2245]
### Troubleshooting docs

- [X] My problem is not solved in the Troubleshooting docs

### Latest version of Mamba

- [X] My problem is not solved with the latest version

### Tried in Conda?

Not reproducible with Conda

### Describe your issue

How can I use micromamba when I'm trying to activate an env and run something from SSH?

with conda I can do something `ssh my-vm bash -c "source ~/miniconda3/bin/activate test && which python"` and that points to the right thing. But there's no `activate` in micromamba (is there?) so how can I use it in a script and/or from SSH?

**Comment by jonashaag** [comment]
Hi!

> But there's no activate in micromamba (is there?)

Why don’t you look this up before creating an issue here?

**Comment by mil-ad** [comment]
> Why don’t you look this up before creating an issue here?

I did, there's the `activate` switch but how can I `source `that?

**Comment by jonashaag** [comment]
Same way as Conda. I think `source` is deprecated. For one-off commands you might want to use `run` instead

**Comment by mil-ad** [comment]
didn't know about `run`. Will try it out, thanks

**Comment by jonashaag** [comment]
I tend to avoid activation stuff as far as possible. It causes a lot of confusion

**Comment by mil-ad** [comment]
agreed. The whole subshelling thing is also quite confusing when you're trying to automate something in a script or do things remotely over ssh. Would be great to have one or two examples in the docs.

**Comment by mil-ad** [comment]
`micromamba run` works great!!

**Comment by jonashaag** [comment]
> Would be great to have one or two examples in the docs.

Feel free to come up concrete questions, and even better, responses to the questions. I'm happy to add them to the docs. In general I think the micromamba activation is supposed to work identical/very similar to the Conda one. I'm not sure the Conda activation is well documented though.

**Comment by wesleykendall** [comment]
I do have a question that would have helped me in the docs, and I think it's related enough to this convo to bring it up here and see if anyone has a quick answer (happy to open a new issue though)

`micromamba run CMD` seems fundamentally different than `micromamba activate && CMD && micromamba deactivate`.

The former is much more heavyweight. For example, try running:

```
docker run -it mambaorg/micromamba:1.2.0 /bin/bash -c 'time micromamba run -n base echo "hi"'
```

Even on a new Apple M2 chip, the overhead is 0.22 seconds. It sounds small, but this adds up when I chain together a workflow of 100 short tasks that run commands in a variety of pre-built envs. This is especially true on older hardware that has a 0.6 second overhead. In my use case, the overhead of `micromamba run` takes up half the time of my workflow.

My question - How can I re-create what `activate && CMD && deactivate` does in a python subprocess?

I checked the output of `micromamba shell hook --shell bash`. It seems like it is:

1. Setting up aliases for activate/deactivate to work
2. Setting up other functions to keep track of the stack of envs
3. Setting up cmd completion
4. Eventually calling `micromamba shell --shell` in `__mamba_activate`

It seems like 4) is the core thing I need to somehow call since the shell stuff is irrelevant to what I'm trying to do. I'm having a hard time digging into exactly what this code is doing (sorry my C++ is rusty!)

TLDR - Is there a more lightweight way I can `micromamba run` in a subprocess? I assume it's not as trivial as just setting `PATH` to point to the env bin folder

**Comment by jonashaag** [comment]
Interesting! Under the hood `run` generates a bash script and runs `activate` and then your command. So my guess it that running this script has a lot of overhead because your bash loads slowly, or because creating the file is slow. Could you please benchmark how long it takes to simply run an empty bash in your setup?

**Comment by jonashaag** [comment]
Any news here?

**Comment by mforbes** [comment]
@wesleykendall What is this `micromamba deactivate` you mention?  There is no mention of this subcommand in [the documentation](https://mamba.readthedocs.io/en/latest/user_guide/concepts.html#activation-deactivation) or in the `micromamba --help`.  (I know about `conda deactivate`... trying to see if there is a documentation or other bug here).

**Comment by jonashaag** [comment]
There is also

```
micromamba shell [-n ...]
```

but it's pretty much the same as `micromamba run [-n ...] $SHELL`

**Comment by mforbes** [comment]
Is that what `shell` does???

```bash
$ micromamba shell --help
Generate shell init scripts
...
```

I would never have guessed from the description!  I thought it modified `~/.bashrc` or something.  Thanks!  (Documentation really needs updating.)

**Comment by jonashaag** [comment]
Yeah, please make improvements to the docs!

See also https://github.com/mamba-org/mamba/issues/2072

**Comment by jonashaag** [comment]
It actually does two things, `micromamba shell` spawns a new one-off shell and `micromamba shell something` does the more "permanent" stuff like changing your bashrc

I think the `deactivate` isn't documented in `--help` because it's not implemented in the `micromamba` binary but in a shell function. But we should document it there anyways, I think we already have documentation for `activate` there?

**Comment by mforbes** [comment]
I can try to improve the docs, but I am not sure I understand what `micromamba shell -p <env>` is trying to do.

It does spawn a shell, but that shell does not seem to activate the specified environment, and also sources the users initialization files.  In my case, this activates another conda environment I generally use, which is very confusing.  I would have thought that it would spawn a shell with the specified environment active.  It seems like this was changed in #2538, though I don't know why yet.  Is this a bug?

In any case, I think that for `micromamba shell` to be useful generally, there needs to be a way to suppressing startup files like `micromamba run -p <env> bash --init-file <init_file>`, which works because`run` uses a script as you mentioned above, and hence does not load any init files.

**Comment by jonashaag** [comment]
You might want to change your init scripts to only activate the default environment if no other environment is already activated. If that's not in the default code shipped by Mamba we should change it.

**Comment by jonashaag** [comment]
Personally I've stopped using activation in the init scripts entirely because it causes a lot of pain

**Comment by mforbes** [comment]
My aim is to make a`Makefile` that allows users to robustly provision an environment, even if they have chosen poorly to use activation:-)

**Comment by jonashaag** [comment]
I don't think it's possible to do so if you mean robust = can have preexisting mamba installs and potentially custom init scripts. 

If you can reproduce the shell problem in a "vanilla" environment that's definitely a bug!

**Comment by jonashaag** [comment]
Closing due to inactivity
